zigzag-circuit.md

```diff
@@ -0,0 +1,122 @@
+
+- `LAYERS : UInt`: Number of layers
+- `LAYER_CHALLENGES[0..LAYERS]`: Number of challenges per layer
+- `EXPANSION_DEGREE: UInt`: Degree of the bipartite expander graph to extend dependencies between layers
+- `TREE_DEPTH: UInt`: Height of the Merkle tree
+
+**Public Inputs**: *Inputs that the prover uses to generate a SNARK proof and that the verifier uses to verify it*
+
+- `ReplicaId : Fr`: A unique identifier for the replica Id
+- `CommD : Fr`: the Merkle tree root hash of the original data (input to the first layer)
+- `CommRLast : Fr`: the Merkle tree root hash of the output of the last layer 
+- `CommRStar : Fr`: the aggregate of each layer's Merkle tree root hash
+- Inclusion paths: Binary representation of the Merkle tree path that must be proven packed into a single `Fr` element. We have the following inclusion paths:
+  - `InclusionPaths_{i=0..LAYERS}_{0..LAYER_CHALLENGES[i]}`: At each layer `i` we have `LAYER_CHALLENGES[i]` inclusion paths.
+  - `ParentsInclusionPaths_{l=0..LAYERS}_{c=0..LAYER_CHALLENGES[i]}_{0..EXPANSION_DEGREE+ BASE_DEGREE}`: At each layer `l` we have `LAYER_CHALLENGES[i]` an inclusion path for each parent node of the corresponding `InclusionPaths_{l}_{c}`.
+
+**Private Inputs**: *Inputs that the prover uses to generate a SNARK proof, these are not needed by the verifier to verify the proof*
+
+- `CommR_{i=0..LAYERS-1}`: Commitment of the the encoded data at layer `i`. 
+- Inclusion Proof: For each inclusion path in the public inputs, we provide a Merkle Tree path
+  - `InclusionHash_{i=0..LAYERS}_{0..LAYER_CHALLENGES[i]}_{0..TREE_DEPTH-1}`
+  - `ReplicaInclusionHash_{i=0..LAYERS}_{0..LAYER_CHALLENGES[i]}_{0..TREE_DEPTH-1}`
+  - `ParentInclusionHash_{l=0..LAYERS}_{c=0..LAYER_CHALLENGES[i]}_{0..EXPANSION_DEGREE+ BASE_DEGREE}_{0..TREE_DEPTH-1}`
+  - `InclusionLeaf_{i=0..LAYERS}_{0..LAYER_CHALLENGES[i]}`
+  - `ReplicaInclusionLeaf_{i=0..LAYERS}_{0..LAYER_CHALLENGES[i]}`
+  - `ParentInclusionLeaf_{l=0..LAYERS}_{c=0..LAYER_CHALLENGES[i]}_{0..EXPANSION_DEGREE+ BASE_DEGREE}`
+
+**Circuit:**
+
+- **Check** `ReplicaId` is equal to its bit representation
+
+  ```
+  Assign replica_id_bits = Fr_to_bits(ReplicaId)
+  Check Packed(replica_id_bits) == ReplicaId
+  ```
+
+- For each `l = 0..LAYERS`:
+
+  - For each `c = 0..LAYERS_CHALLENGES[l]`
+
+    - Inclusion checks:
+
+      - Correct inclusions proofs: **Check**  that all the inclusion proofs are correct
+
+        ```
+        Check MerkleTreeVerify(InclusionHash_{l}_{c}_{0..TREE_DEPTH-1})
+        Check MerkleTreeVerify(ReplicaInclusionHash_{l}_{c}_{0..TREE_DEPTH-1})
+        
+        For p = 0..(EXPANSION_DEGREE + BASE_DEGREE):
+        	Check MerkleTreeVerify(ParentInclusionHash_{l}_{c}_{p}_{0..TREE_DEPTH-1})
+        ```
+
+      - Correct layer: **Check** that `CommR_{l}` is matching the Replica Inclusion proofs root hash and `CommR_{l-1}` is matching the Inclusion proof root hash. (If `l=0`, use `CommD` instead, if `l=LAYER-1` use `CommRLast` instead).
```

Comment:
I get the clarification about the `0` and `LAYER-1` corner cases but I think it's a bit ambiguous how it's phrased because we replace `CommD` only in the `CommR_{l-1}` line and `CommRLast` in the `CommR_{l}` line (if I'm following this correctly).

Reply:
that's right.. I will make it explicit